File: Ozan/day_2.java
// Solution for 74. Search a 2D Matrix
// link: https://leetcode.com/problems/search-a-2d-matrix/description/

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int numberOfRows = matrix.length;
        int numberOfColumns = matrix[0].length;
        
        // 1. find row
        int first = 0;
        int last = numberOfRows -1;
        int rowIndex = (last + first)/2;

        while (first<=last) {
            if (matrix[rowIndex][0]<=target && target<=matrix[rowIndex][numberOfColumns-1])
                break;

            if (matrix[rowIndex][0]>target) {
                last = rowIndex - 1;
                rowIndex = (first+last)/2;
            } else if (matrix[rowIndex][0]<target) {
                first = rowIndex + 1;
                rowIndex = (first+last)/2;
            } 
        }

        // 2. find column
        first = 0;
        last = matrix[0].length-1;
        int indexOfMiddle = (last + first)/2;
        
        boolean res = false;
        while (first<=last) {
            if (matrix[rowIndex][indexOfMiddle]==target) {
                res = true;
                break;
            }

            if (matrix[rowIndex][indexOfMiddle]>target) {
                last = indexOfMiddle - 1;
                indexOfMiddle = (first+last)/2;
            } else if (matrix[rowIndex][indexOfMiddle]<target) {
                first = indexOfMiddle + 1;
                indexOfMiddle = (first+last)/2;
            }
        }
        
        return res;
    }
}

File: Deniz/day_2.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix or not matrix[0]:
            return False
        m, n = len(matrix), len(matrix[0])
        return self.binarySearch(matrix, target, 0, m * n - 1, n)
    
    def binarySearch(self, matrix, target, left, right, n):
        if left > right:
            return False
        
        mid = (left + right) // 2
        mid_value = matrix[mid // n][mid % n]
        
        if target == mid_value:
            return True
        elif target < mid_value:
            return self.binarySearch(matrix, target, left, mid - 1, n)
        else:
            return self.binarySearch(matrix, target, mid + 1, right, n)